wolfcrypt/src/port/Renesas/renesas_tsip_util.c

```diff
@@ -0,0 +1,654 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+#include <wolfssl/wolfcrypt/settings.h>
+
+#if defined(WOLFSSL_RENESAS_TSIP)
+
+#include <wolfssl/wolfcrypt/wc_port.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+
+#include <wolfssl/wolfcrypt/port/Renesas/renesas-tsip-crypt.h>
+#include <wolfssl/wolfcrypt/memory.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+#include <wolfssl/wolfcrypt/aes.h>
+#include <wolfssl/ssl.h>
+#include <wolfssl/internal.h>
+
+#include <stdio.h>
+/* mutex */
+wolfSSL_Mutex tsip_mutex;
+static int tsip_CryptHwMutexInit_ = 0;
+
+/* ./ca-cert.der.sign,  */
+/* expect to have these valiables defined at user application */
+extern uint32_t s_flash[];
+extern uint32_t s_inst1[R_TSIP_SINST_WORD_SIZE];
+extern uint32_t s_inst2[R_TSIP_SINST2_WORD_SIZE];
+static const byte *ca_cert_sig;
+
+/* user key */
+static tsip_key_data g_user_key_info;
+/* tsip only keep one encrypted ca public key */
+#if defined(WOLFSSL_RENESAS_TSIP_TLS)
+static uint32_t g_encrypted_publicCA_key[R_TSIP_SINST_WORD_SIZE];
+static uint32_t g_CAscm_Idx;          /* index of CM table    */
+#endif
+
+static int tsip_CryptHwMutexInit(wolfSSL_Mutex* mutex) {
+    return wc_InitMutex(mutex);
+}
+
+static int tsip_CryptHwMutexLock(wolfSSL_Mutex* mutex) {
+    return wc_LockMutex(mutex);
+}
+
+static int tsip_CryptHwMutexUnLock(wolfSSL_Mutex* mutex) {
+    return wc_UnLockMutex(mutex);
+}
+
+/*
+* lock hw engine.
+* this should be called before using engine.
+*/
+int tsip_hw_lock()
+{
+    int ret = 0;
+
+    WOLFSSL_MSG("enter esp_sha_hw_lock");
+
+    if(tsip_CryptHwMutexInit_ == 0){
+        ret = tsip_CryptHwMutexInit(&tsip_mutex);
+        if(ret == 0) {
+            tsip_CryptHwMutexInit_ = 1;
+        } else {
+            WOLFSSL_MSG(" mutex initialization failed.");
+            return -1;
+        }
+    }
+    if(tsip_CryptHwMutexLock(&tsip_mutex) != 0) {
+        /* this should not happens */
+        return -1;
+    }
+
+    WOLFSSL_MSG("leave tsip_sha_try_hw_lock");
+    return ret;
+}
+
+/*
+* release hw engine
+*/
+void tsip_hw_unlock( void )
+{
+    WOLFSSL_MSG("enter tsip_hw_unlock");
+    /* unlock hw engine for next use */
+    tsip_CryptHwMutexUnLock(&tsip_mutex);
+    WOLFSSL_MSG("leave tsip_hw_unlock");
+}
+/* check if tsip tls functions can be used for the cipher      */
+/* cipher0 : in the some cipher suite,                         */
+/*           first byte becomes greater than 0, otherwise 0x00 */
+/* side    : CLIENT END or SEVER END                           */
+byte tsip_useable(byte cipher0, byte cipher, byte side)
+{
+    if (cipher0 > 0x00) return 0;
+
+    if ((cipher == l_TLS_RSA_WITH_AES_128_CBC_SHA ||
+         cipher == l_TLS_RSA_WITH_AES_128_CBC_SHA256 ||
+         cipher == l_TLS_RSA_WITH_AES_256_CBC_SHA ||
+         cipher == l_TLS_RSA_WITH_AES_256_CBC_SHA256) &&
+         side == WOLFSSL_CLIENT_END)
```

Comment:
Is TSIP only usable on the client side?  If so, please double check this is in the documentation.

Reply:
We can use certificate verification on both sides. But , common crypt of TSIP TSL-linked capability is only for client-side on current TSIP version.
I'll check it in the document.

Reply:
Reminder to add this to "Limitations" section of docs.  I didn't see it there yet, thanks.

Reply:
I will! And I have reconstructed tsip_usable() to check if encrypted rsa key index has been generated by TSIP.  If not, we cannot use TSIP.